File: multi_task/test_tasks.py
import decimal
import argparse
import time
import sys
def is_prime(n):
    if n <= 1: return False
    for i in range(2, int(n**0.5)+1):
        if n % i == 0: return False
    return True

def calc_pi(precision):
    decimal.getcontext().prec = precision + 2
    C = 426880 * decimal.Decimal(10005).sqrt()
    M = 1
    L = 13591409
    X = 1
    K = 6
    S = L
    for i in range(1, precision//14 + 2):
        M = (K**3 - 16*K) * M // (i+1)**3
        L += 545140134
        X *= -262537412640768000
        S += decimal.Decimal(M * L) / X
        K += 12
    return C / S


def fib(n):
    if n <= 1: return n
    return fib(n-1) + fib(n-2)



import numpy as np

def matrix_ops(size):
    a = np.random.rand(size, size)
    b = np.random.rand(size, size)
    return np.linalg.det(a @ b)





def monte_carlo(samples):
    inside = 0
    for _ in range(samples):
        x, y = np.random.random(), np.random.random()
        if x**2 + y**2 <= 1:
            inside += 1
    return 4 * inside / samples



# 测试大数（示例参数可能需要调整）
# sum(is_prime(n) for n in range(1_000_000, 6_000_001))

# 示例：计算1000位圆周率（耗时约2-5分钟）
# calc_pi(25_000)

# 示例：2000x2000矩阵（需约2GB内存，耗时约3-5分钟）
# matrix_ops(2000)

# monte_carlo(30_000_000)

# 测试参数建议：n=35-40（指数级复杂度）
# fib(41)

def main():
    parser = argparse.ArgumentParser(description='数学计算任务执行器')
    subparsers = parser.add_subparsers(dest='task', required=True, help='任务类型')

    # 质数检测
    prime_parser = subparsers.add_parser('prime', help='质数检测')
    prime_parser.add_argument('start', type=int, help='起始值')
    prime_parser.add_argument('end', type=int, help='结束值')

    # 圆周率计算
    pi_parser = subparsers.add_parser('pi', help='计算圆周率')
    pi_parser.add_argument('precision', type=int, help='精度位数')

    # 斐波那契数列
    fib_parser = subparsers.add_parser('fib', help='斐波那契数列')
    fib_parser.add_argument('n', type=int, help='项数')

    # 矩阵运算
    matrix_parser = subparsers.add_parser('matrix', help='矩阵运算')
    matrix_parser.add_argument('size', type=int, help='矩阵维度')

    # 蒙特卡洛
    mc_parser = subparsers.add_parser('mc', help='蒙特卡洛积分')
    mc_parser.add_argument('samples', type=int, help='采样次数')

    args = parser.parse_args()

    try:
        start_time = time.time()
        if args.task == 'prime':
            count = sum(1 for n in range(args.start, args.end + 1) if is_prime(n))
            print(f"发现 {count} 个质数")

        elif args.task == 'pi':
            pi = calc_pi(args.precision)
            print(f"前50位: {str(pi)[:50]}...")

        elif args.task == 'fib':
            result = fib(args.n)
            print(f"fib({args.n}) = {result}")

        elif args.task == 'matrix':
            det = matrix_ops(args.size)
            print(f"行列式值: {det:.4e}")

        elif args.task == 'mc':
            pi_approx = monte_carlo(args.samples)
            print(f"π ≈ {pi_approx}")

        print(f"耗时 {time.time() - start_time:.2f}秒")
    except Exception as e:
        print(f"执行出错: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    main()

# # 质数检测
# python test_tasks.py prime 1000000 6000001
#
# # 计算圆周率
# python test_tasks.py pi 25000
#
# # 斐波那契
# python test_tasks.py fib 41
#
# # 矩阵运算（2000x2000）
# python test_tasks.py matrix 2000
#
# # 蒙特卡洛
# python test_tasks.py mc 30000000

File: multi_task/test_c++/test_tasks.cpp
#include <iostream>
#include <cmath>
#include <random>
#include <chrono>
#include <vector>
#include <iomanip>
using namespace std;

// 判断质数
bool is_prime(int n) {
    if (n <= 1) return false;
    for (int i = 2; i <= std::sqrt(n); ++i) {
        if (n % i == 0) return false;
    }
    return true;
}

// 计算圆周率
// 修复后的圆周率计算函数
//mpf_class calc_pi(int precision) {
//    mpf_set_default_prec(precision + 2);
//    mpf_class C = 426880 * sqrt(mpf_class(10005));
//    mpf_class M = 1;
//    mpf_class L = 13591409;
//    mpf_class X = 1;
//    mpf_class K = 6;
//    mpf_class S = L;
//
//    for (int i = 1; i <= precision / 14 + 2; ++i) {
//        M = (K * K * K - 16 * K) * M / ((i + 1) * (i + 1) * (i + 1));
//        L += 545140134;
//        X *= mpf_class(-262537412640768000);
//        S += M * L / X;
//        K += 12;
//    }
//
//    return C / S;
//}



long double calculate_pi_series(long long iterations) {
    long double pi = 0.0;
    int sign = 1;
    long long out_count = iterations /10;
    for (long long  i = 0; i < iterations; ++i) {
        long double term = 1.0 / (2*i + 1);
        pi += sign * term;
        sign *= -1; // 交替符号

        if (i % out_count == 0) { // 每百万次输出进度
            cout << "Iteration " << i
                 << ": π ≈ " <<std::setprecision(50) << 4*pi
                 << endl;
        }
    }
    return 4 * pi;
}



// 斐波那契数列
int fib(int n) {
    if (n <= 1) return n;
    return fib(n - 1) + fib(n - 2);
}

// 矩阵运算
double matrix_ops(int size) {
    std::vector<std::vector<double>> a(size, std::vector<double>(size));
    std::vector<std::vector<double>> b(size, std::vector<double>(size));
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);

    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < size; ++j) {
            a[i][j] = dis(gen);
            b[i][j] = dis(gen);
        }
    }

    std::vector<std::vector<double>> c(size, std::vector<double>(size, 0));
    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < size; ++j) {
            for (int k = 0; k < size; ++k) {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }

    double trace = 0;
    for (int i = 0; i < size; ++i) {
        trace += c[i][i];
    }
    return trace;
}

// 蒙特卡洛积分
double monte_carlo(int samples) {
    int inside = 0;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);

    for (int i = 0; i < samples; ++i) {
        double x = dis(gen);
        double y = dis(gen);
        if (x * x + y * y <= 1) {
            inside++;
        }
    }
    return 4.0 * inside / samples;
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <task> <args...>" << std::endl;
        return 1;
    }

    std::string task = argv[1];
    auto start_time = std::chrono::high_resolution_clock::now();

    try {
        if (task == "prime") {
            int start = std::stoi(argv[2]);
            int end = std::stoi(argv[3]);
            int count = 0;
            for (int n = start; n <= end; ++n) {
                if (is_prime(n)) count++;
            }
            std::cout << "发现 " << count << " 个质数" << std::endl;
        } else if (task == "pi") {
            long long precision = std::stoll(argv[2]);
            long double pi =  calculate_pi_series(precision);
            std::cout << std::setprecision(50) << pi; // 输出50位有效数字
            std::cout << "..." << std::endl;
        } else if (task == "fib") {
            int n = std::stoi(argv[2]);
            int result = fib(n);
            std::cout << "fib(" << n << ") = " << result << std::endl;
        } else if (task == "matrix") {
            int size = std::stoi(argv[2]);
            double det = matrix_ops(size);
            std::cout << "行列式值: " << std::scientific << std::setprecision(4) << det << std::endl;
        } else if (task == "mc") {
            int samples = std::stoi(argv[2]);
            double pi_approx = monte_carlo(samples);
            std::cout << "π ≈ " << pi_approx << std::endl;
        } else {
            std::cerr << "未知任务: " << task << std::endl;
            return 1;
        }

        auto end_time = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = end_time - start_time;
        std::cout << "耗时 " << elapsed.count() << " 秒" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "执行出错: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

//test_tasks mc 400000000
//test_tasks fib 48
//test_tasks prime 1000000 6000001
//test_tasks pi 5000000000